### ISSUE TITLE
Clang diagnostics: false-positive warning "unused lambda-capture"

### ISSUE THREAD

**Issue by pprettysimpple** [llvm/llvm-project#134797]
In this example clang (trunk) emits warning that lambda capture is not needed, when actually code does not compile without it.
https://godbolt.org/z/jGM18c93e

```cpp
#include <iostream>

struct context {
    void do_smth() {
        auto l = [this](auto x) { this->act(x); };
        l(42);
        l(43.0);
    }

    static void act(int x) {
        std::cout << x;
    }
    void act(double x) {
        std::cout << x;
    }
};

int main() {
    context c;
    c.do_smth();
}

```
